<a href="https://colab.research.google.com/github/rosslogan702/hypothesis_testing_notes/blob/master/two_sample_t_test.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Hypothesis Testing - 2 Sample T-Test

# Contents

The focus of this notebook is the 2 Sample T-Test.

The notebook will cover the following:



1.   Description
2.   Manual Calculation
3.   Practical Examples using Scipy Library
4. Assumptions



# 1. Description

A 2 sample t-test investigates whether the means of two independent samples of data differ from each other.

### Example

Suppose that the average sales per day last week for an online store was 210 items per day. 

This week the average sales figure per day is 285 items. Did the average sales figure per day change or is this just part of natural fluctuations?

The 2 Sample T-Test could be used to try and determine this.



# 2. Manual Calculation

The following will detail the manual calculation steps required for performing a 2 Sample T-Test. The steps will be detailed through an example.



## Manual Example - Voter Ages for different states

We want to investigate whether the voting age for the state of florida and the state of new york is different.

Two random samples of the ages of voters are collected. The samples of ages for the state of florida and new york are shown below.

We want to test at the significance level of 5%.

In [0]:
import numpy as np
florida_voter_ages = [44, 47, 60, 65, 47, 52, 56, 56, 44, 48, 50, 47, 55, 49, 
                      50, 50, 50, 41, 53, 47, 39, 50, 52, 43, 53, 51, 59, 53, 
                      51, 46, 32, 38, 32, 29, 40, 33, 30, 32, 29, 27, 30, 36, 
                      30, 31, 34, 27, 34, 26, 29, 33]

ny_voter_ages = [46, 42, 39, 57, 45, 48, 46, 50, 41, 50, 45, 50, 51, 44, 50, 
                 46, 52, 57, 47, 53, 50, 48, 48, 41, 52, 40, 48, 44, 49, 51, 
                 27, 26, 30, 25, 31, 29, 28, 31, 33, 24, 23, 27, 23, 28, 27, 
                 31, 29, 26, 27, 25]

## Step 1 - Define Null & Alternative Hypothesis

$H_0$: There difference between the sample mean of florida voter ages and ny voter ages is equal to 0.

$H_0: \mu_\text{florida voter age} - \mu_\text{new york voter age} = 0$

$H_A$: The diference between the sample mean of florida voter ages and ny voter ages is not equal to 0.

$H_A: \mu_\text{florida voter age} - \mu_\text{new york voter age} \neq 0$

## Step 2 - Calculate the T-Statistic

  

$t\text{-}statistic = \frac{\bar{x_1} - \bar{x_2}}{\sqrt{\dfrac{s_1^2}{n_1} + \dfrac{s_2^2}{n_2}}}$ 

This formula makes no assumption about the variance of the samples, if they are the same then it can be simplified.

variable meanings:  

$\bar{x_1}$ = sample mean of florida voter ages  
$\bar{x_2}$ = sample mean of new york voter ages   
$\bar{s_1^2}$ = variance of florida voter ages sample  
$\bar{s_2^2}$ = variance of new york voter ages sample  
$n_1$ = number of samples in florida voter ages  
$n_2$ = number of samples in new york voter ages 

In [0]:
x_1 = np.mean(florida_voter_ages)
x_2 = np.mean(ny_voter_ages)

s_1 = np.std(florida_voter_ages)
s_2 = np.std(ny_voter_ages)

n_1 = len(florida_voter_ages)
n_2 = len(ny_voter_ages)

print('x_1: {0:.2f}'.format(x_1))
print('x_2: {0:.2f}'.format(x_2))
print('florida_voter_age_std: {0:.2f}'.format(s_1))
print('ny_voter_age_std: {0:.2f}'.format(s_2))
print('florida_voter_num_samples: {0:.2f}'.format(n_1))
print('ny_voter_num_samples: {0:.2f}'.format(n_2))

x_1: 42.80
x_2: 39.60
florida_voter_age_std: 10.34
ny_voter_age_std: 10.61
florida_voter_num_samples: 50.00
ny_voter_num_samples: 50.00


In [0]:
from math import sqrt
t_stat = (x_1 - x_2)/np.sqrt((s_1**2/n_1) + (s_2**2/n_2))
print('t-statistic: {0:.3f}'.format(t_stat))

t-statistic: 1.527


## Step 3 Find the T-Table Value

To find the t-table value you need:

The alpha level: given as 5% in the question  
The degrees of freedom - which is the total number of samples from each -2     
A T-Table - http://www.sjsu.edu/faculty/gerstman/StatPrimer/t-table.pdf  

Looking up 98 degrees of freedom in the left column and the corresponding 0.05 in the top row gives a value of 1.990.

(98 degrees of freedom did not exist in the t-table used so we drop to the next closest lower degress of freedom which was 80).

This value means that we would expect most values to fall under 1.990. If our calculated t-value falls within this range, the null hypothesis is likely true.



## Step 4 - Compare the calculated t-value to the critical value from t-table

In this case the value from step 2 does not fall into the range calculated in step 3, so we do not reject the null hypothesis.

The value 1.527 falls within this region.

(When inspecting the t-stat value from section 3 on the same dataset, the values are slightly different however this is most likely due to the scipy library returning a more accurate t-table value, we have done an approximation using 80 DOF when we were really interested in 98).

# 3. Practical Examples using Scipy Library

Scipy code: https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html

## Example 1 - Audio Book App Usage Time

A new audio book app collects data from two different weeks. The data is the app usage time of its users. It finds that the average amount of listening time from its user in the first week was ~ 25.4 minutes and ~ 29 minutes in the second week.

Did the average usage time per user change or is this part of natural fluctuations?

In [0]:
# Times in minutes of audio book app usage listening time from week 1 in minutes
week_1_times = [23.90506824, 26.67631982, 27.27433886, 24.25757125, 
                32.40423483, 39.56919357, 23.07010059, 29.82068109,
                27.59433809, 28.05639569, 27.06757262, 30.41192979,
                25.71358554, 24.94294823, 28.23123807, 24.95337555,  
                18.51231639, 27.46234762, 28.38016611, 13.91205901,
                29.02615866, 26.90746774, 22.8677726,  24.8938289,  
                25.96947935, 26.86869621, 20.72676456, 27.35988314,
                20.68408581, 21.19846143, 16.25800931, 23.92517681,  
                24.47923229, 29.47050863, 27.28425372, 26.93339272,  
                28.61026924, 18.88377042, 33.65468651, 25.69470077,
                20.98291356, 22.69700387, 28.60278855, 21.36000443,  
                30.77685156, 20.83415999, 23.79367158, 19.7556718,
                29.54421084, 20.1433138]

In [0]:
# Times in mintes of audio book app usage listening time from week 2 in minutes
week_2_times = [18.63431907, 31.28788036, 34.96797943, 21.81678117, 
                28.21619974, 39.39313736, 35.52223207, 27.54222109, 
                33.64395433, 25.31673581, 28.81392191, 30.7358016, 
                26.37241881, 26.0945555, 26.34073477, 19.42196017, 
                32.58797652, 24.84001926, 28.93348335, 20.43667584, 
                22.72495967, 32.31728012, 35.384306, 29.66709637, 
                24.53512973, 30.91406007, 19.56117513, 24.90816833, 
                30.13163726, 31.47466199, 27.77683598, 16.51307462, 
                35.0770162, 31.74818107, 36.36053496, 27.70500593, 
                29.49869936, 27.65575346, 37.18504075, 25.16055104, 
                29.26553553, 38.22163057, 28.92102091, 24.8215439, 
                38.30155495, 34.76020645, 22.26869162, 28.82593733, 
                32.00975127, 36.46437665]

## Step 1 - Define Null Hypothesis and Alternative Hypothesis

Null Hypothesis = $H_0$  
Alternative Hypothesis = $H_A$  

$H_0$: The difference between the sample mean of app listening times for week 1 and week 2 is equal to 0.

$H_0: \mu_\text{week_1} - \mu_\text{week_2} = 0$

$H_A$: The difference between the sample mean of app listening times for week 1 and week 2 is not equal to 0.

$H_A: \mu_\text{week_1} - \mu_\text{week_2} \neq 0$

## Step 2 - Prepare Data & Run Test

In [0]:
from scipy.stats import ttest_ind
import numpy as np

In [0]:
# Collect Statistics about each week
week_1_mean = np.mean(week_1_times)
week_2_mean = np.mean(week_2_times)

week_1_std = np.std(week_1_times)
week_2_std = np.std(week_2_times)

In [0]:
# Print statistics
print('Week 1 mean: {}'.format(week_1_mean))
print('Week 1 std:  {}'.format(week_1_std))
print('Week 2 mean: {}'.format(week_2_mean))
print('Week 2 std:  {}'.format(week_2_std))

Week 1 mean: 25.4480593952
Week 1 std:  4.531693386680561
Week 2 mean: 29.0215681076
Week 2 std:  5.497966708987187


In [0]:
# Run test using scipy lib
_, pval = ttest_ind(week_1_times, week_2_times)

## Step 3 - Collect & Analyse Results

In [0]:
print('pval: {:.6f}'.format(pval))
if pval < 0.05:
  print("Result is statistically significant! The listening time between week 1 and week 2 has changed!")
else:
  print("Result is not statistically signifcant! The listening time between week 1 and week 2 has not changed!")

pval: 0.000677
Result is statistically significant! The listening time between week 1 and week 2 has changed!


## Example 2 - Voter ages for different states.

We want to investigate whether the voting age for the state of florida and the state of new york is different.

Two random samples of the ages of voters are collected. The samples of ages for the state of florida and new york are shown below.



In [0]:
florida_voter_ages = [44, 47, 60, 65, 47, 52, 56, 56, 44, 48, 50, 47, 55, 49, 
                      50, 50, 50, 41, 53, 47, 39, 50, 52, 43, 53, 51, 59, 53, 
                      51, 46, 32, 38, 32, 29, 40, 33, 30, 32, 29, 27, 30, 36, 
                      30, 31, 34, 27, 34, 26, 29, 33]

In [0]:
ny_voter_ages = [46, 42, 39, 57, 45, 48, 46, 50, 41, 50, 45, 50, 51, 44, 50, 
                 46, 52, 57, 47, 53, 50, 48, 48, 41, 52, 40, 48, 44, 49, 51, 
                 27, 26, 30, 25, 31, 29, 28, 31, 33, 24, 23, 27, 23, 28, 27, 
                 31, 29, 26, 27, 25]

## 1. Define Null Hypothesis and Alternative Hypothesis

$H_0$: There difference between the sample mean of florida voter ages and ny voter ages is equal to 0.

$H_0: \mu_\text{florida voter age} - \mu_\text{new york voter age} = 0$

$H_A$: The diference between the sample mean of florida voter ages and ny voter ages is not equal to 0.

$H_A: \mu_\text{florida voter age} - \mu_\text{new york voter age} \neq 0$



## Step 2 - Prepare Data & Run Test

In [0]:
from scipy.stats import ttest_ind
import numpy as np

In [0]:
# Collect Statistics about each week
florida_voter_mean = np.mean(florida_voter_ages)
ny_voter_mean = np.mean(ny_voter_ages)

florida_voter_std = np.std(florida_voter_ages)
ny_voter_std = np.std(ny_voter_ages)

In [0]:
# Print statistics
print('Florida Voter Mean: {}'.format(florida_voter_mean))
print('Florida Voter std:  {}'.format(florida_voter_std))
print('New York Mean: {}'.format(ny_voter_mean))
print('New York std:  {}'.format(ny_voter_std))

Florida Voter Mean: 42.8
Florida Voter std:  10.342146779078316
New York Mean: 39.6
New York std:  10.607544484940895


In [0]:
# Run test using scipy lib
t_stat, p_val = ttest_ind(florida_voter_ages, ny_voter_ages)

In [0]:
print('t_stat: {0:.3f}'.format(t_stat))
print('p_val: {0:.3f}'.format(p_val))

t_stat: 1.512
p_val: 0.134


## Step 3 - Collect & Analyse Results

In [0]:
print('pval: {:.6f}'.format(pval))
if pval < 0.05:
  print("Result is statistically significant!")
else:
  print("Result is not statistically signifcant!")

pval: 0.000677
Result is statistically significant!


# 4. Assumptions & Notes

The order of the datasets entered into the scipy library does not matter. We are testing whether there is a significant difference between both sets of data. The order is irrelevant.  

The data is assumed to be approximately normally distributed across both sets of data.
